File: security_cam/integration_test/screen_off_gate_test.dart
import 'dart:io';

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:integration_test/integration_test.dart';
import 'package:shared_preferences/shared_preferences.dart';
import 'package:sqflite/sqflite.dart' as sql;
import 'package:sqflite_common_ffi/sqflite_ffi.dart';

import 'package:security_cam/sensors/permissions_service.dart';
import 'package:security_cam/state/monitor_controller.dart';
import 'package:security_cam/storage/event_log.dart';
import 'package:security_cam/storage/settings_store.dart';
import 'package:security_cam/storage/snapshot_store.dart';
import 'package:security_cam/ui/app.dart';

/// Screen-off continuity gate: while the FGS runs with `camera|microphone`,
/// the camera analysis + event pipeline must keep working with the screen off.
///
/// The host runner (`tool/run_android_integration_tests.sh`) watches the
/// `[itest] SCREEN_OFF_READY` marker and drives: KEYCODE_POWER (off) ~5s later,
/// then KEYCODE_POWER (on) when `SCREEN_OFF_DONE` is seen. This test asserts
/// monitoring stays healthy through the off window and that a second motion
/// event arrives after the display returns (camera stream survived + resumed).
///
/// Runs with `recordVideo=false` (gateSettings below): the software AVC encoder
/// behind the bound video use case starves the emulator's camera on swiftshader,
/// collapsing analysis to <1 fps within a couple of minutes — while on a real
/// device it's fine, on the headless AOSP image it makes the long gate
/// unreliable. This gate isolates *camera-analysis continuity*; clip recording,
/// export, and resolution are covered by `monitoring_on_device_test.dart`.
void main() {
  final binding = IntegrationTestWidgetsFlutterBinding.ensureInitialized();

  testWidgets('monitoring survives a screen-off window', (tester) async {
    sqfliteFfiInit();
    sql.databaseFactory = databaseFactoryFfi;
    SharedPreferences.setMockInitialValues({});
    final settingsStore = await SettingsStore.open();
    await settingsStore
        .save((await settingsStore.load()).copyWith(recordVideo: false));
    final eventLog = await SqliteEventLog.open(inMemoryDatabasePath);
    final snapDir = await Directory.systemTemp.createTemp('scam_itest_soff');
    final snapshotStore = FileSnapshotStore(snapDir.path);
    final controller = MonitorController(
      settingsStore: settingsStore,
      eventRecorder: eventLog,
      snapshotStore: snapshotStore,
      purgeInterval: null,
      permissionsService: const NoopPermissionsService(),
    );
    await controller.init();
    addTearDown(() async {
      await controller.stop();
      await eventLog.close();
      snapDir.deleteSync(recursive: true);
    });
    await binding.setSurfaceSize(const Size(720, 1280));

    await tester.pumpWidget(SecurityCamApp(
      controller: controller,
      eventLoader: () => eventLog.recent(limit: 200),
      snapshotStore: snapshotStore,
    ));
    await tester.pump();

    Future<void> waitForState(MonitorState state) async {
      final deadline = DateTime.now().add(const Duration(minutes: 2));
      while (controller.state != state && DateTime.now().isBefore(deadline)) {
        await tester.pump(const Duration(milliseconds: 500));
      }
      expect(controller.state, state,
          reason: 'controller.error=${controller.error}');
    }

    Future<List<dynamic>> recent() => eventLog.recent();

    await tester.tap(find.byType(FilledButton));
    await waitForState(MonitorState.monitoring);

    // Baseline: first motion event.
    final deadline = DateTime.now().add(const Duration(minutes: 3));
    while ((await recent()).isEmpty && DateTime.now().isBefore(deadline)) {
      await Future<void>.delayed(const Duration(seconds: 2));
    }
    expect((await recent()), isNotEmpty, reason: 'no baseline motion event');
    final baseline = (await recent()).length;
    debugPrint('[itest] SCREEN_OFF_READY');

    // Window during which the host keeps the display off. Assert monitoring
    // stays healthy throughout (no error, state still monitoring).
    var secondEvent = false;
    var windowError = false;
    final windowEnd = DateTime.now().add(const Duration(seconds: 45));
    while (DateTime.now().isBefore(windowEnd)) {
      // Real-time delay, not tester.pump(): while the display is asleep the
      // engine suspends frame production, so pump() blocks indefinitely.
      await Future<void>.delayed(const Duration(seconds: 2));
      if (controller.error != null ||
          controller.state != MonitorState.monitoring) {
        windowError = true;
        break;
      }
      final rows = await recent();
      if (rows.length > baseline) {
        secondEvent = true;
        break;
      }
    }
    debugPrint('[itest] SCREEN_OFF_DONE');

    // Display is back on: the camera must resume full-rate analysis and a
    // second motion event must arrive within the poll window.
    final recoveryDeadline = DateTime.now().add(const Duration(seconds: 90));
    while (!secondEvent && DateTime.now().isBefore(recoveryDeadline)) {
      await Future<void>.delayed(const Duration(seconds: 2));
      final rows = await recent();
      if (rows.length > baseline) secondEvent = true;
    }

    expect(windowError, isFalse,
        reason: 'monitoring stopped/errored during screen-off: '
            '${controller.error}');
    expect(controller.state, MonitorState.monitoring,
        reason: 'monitoring stopped/errored during screen-off: '
            '${controller.error}');
    expect(controller.error, isNull);
    expect(secondEvent, isTrue,
        reason: 'no motion event recorded after the screen-off window '
            '(camera likely stalled)');

    await tester.tap(find.byType(FilledButton));
    await waitForState(MonitorState.idle);
  });
}

File: android/app/src/androidTest/kotlin/io/securitycam/level2/ScreenOffGateTest.kt
package io.securitycam.level2

import android.content.Context
import android.os.SystemClock
import androidx.test.core.app.ApplicationProvider
import androidx.test.ext.junit.runners.AndroidJUnit4
import io.securitycam.level2.core.AppSettings
import io.securitycam.level2.monitor.MonitorState
import io.securitycam.level2.monitor.MonitorViewModel
import io.securitycam.level2.storage.AppDatabase
import io.securitycam.level2.storage.EncryptedSecretStore
import io.securitycam.level2.storage.RecordedEventRow
import io.securitycam.level2.storage.RoomEventLog
import io.securitycam.level2.storage.SettingsStore
import java.time.Instant
import kotlinx.coroutines.runBlocking
import org.junit.After
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith

/**
 * Screen-off continuity gate (native port of
 * `integration_test/screen_off_gate_test.dart`): while the FGS runs with the
 * camera+microphone type, analysis and the event pipeline must keep working
 * with the display off.
 *
 * The host runner watches logcat for `[itest] SCREEN_OFF_READY`, toggles the
 * screen off ~5 s later, and turns it back on when `SCREEN_OFF_DONE` appears.
 *
 * Runs with `recordVideo=false`: the software AVC encoder starves the
 * emulator's camera under swiftshader over long windows. Clip recording is
 * covered by [MonitoringInstrumentedTest].
 */
@RunWith(AndroidJUnit4::class)
class ScreenOffGateTest {

    private val context: Context =
        ApplicationProvider.getApplicationContext<android.app.Application>()
    private lateinit var vm: MonitorViewModel

    @Before
    fun setUp() = runBlocking {
        clearEvents()
        saveSettings { it.copy(recordVideo = false) }
    }

    @After
    fun tearDown() {
        runBlocking {
            if (::vm.isInitialized && vm.state.value != MonitorState.Idle) {
                vm.stop()
            }
            saveSettings { AppSettings.defaults() }
        }
    }

    private suspend fun saveSettings(transform: (AppSettings) -> AppSettings) {
        val store = SettingsStore(context, EncryptedSecretStore(context))
        store.save(transform(store.load()))
    }

    private suspend fun clearEvents() {
        RoomEventLog(AppDatabase.get(context).eventDao())
            .deleteEvents(Instant.now().plusSeconds(3600))
    }

    private suspend fun recent(): List<RecordedEventRow> =
        RoomEventLog(AppDatabase.get(context).eventDao()).recent(200)

    private fun awaitState(target: MonitorState, timeoutMs: Long = 120_000) {
        val deadline = SystemClock.elapsedRealtime() + timeoutMs
        while (vm.state.value != target && SystemClock.elapsedRealtime() < deadline) {
            Thread.sleep(500)
        }
        assertEquals("controller.error=${vm.error.value}", target, vm.state.value)
    }

    @Test
    fun monitoringSurvivesAScreenOffWindow() = runBlocking {
        vm = MonitorViewModel(application = context as android.app.Application)
        vm.start()
        awaitState(MonitorState.Monitoring)

        // Baseline: first motion event.
        val baselineDeadline = SystemClock.elapsedRealtime() + 3 * 60_000L
        var baseline = 0
        while (SystemClock.elapsedRealtime() < baselineDeadline) {
            baseline = recent().size
            if (baseline > 0) break
            Thread.sleep(2_000)
        }
        assertTrue("no baseline motion event", baseline > 0)
        ItestHarness.mark("SCREEN_OFF_READY")

        // Window during which the host keeps the display off (~45 s). Assert
        // monitoring stays healthy throughout; a second event may already land.
        var secondEvent = false
        var windowError = false
        val windowEnd = SystemClock.elapsedRealtime() + 45_000
        while (SystemClock.elapsedRealtime() < windowEnd) {
            Thread.sleep(2_000)
            if (vm.error.value != null || vm.state.value != MonitorState.Monitoring) {
                windowError = true
                break
            }
            if (recent().size > baseline) {
                secondEvent = true
                break
            }
        }
        ItestHarness.mark("SCREEN_OFF_DONE")

        // Display is back on: a second motion event must arrive within the
        // recovery window (camera survived + resumed full-rate analysis).
        val recoveryEnd = SystemClock.elapsedRealtime() + 90_000
        while (!secondEvent && SystemClock.elapsedRealtime() < recoveryEnd) {
            Thread.sleep(2_000)
            if (recent().size > baseline) secondEvent = true
        }

        assertTrue(
            "monitoring stopped/errored during screen-off: ${vm.error.value}",
            !windowError,
        )
        assertEquals(
            "monitoring stopped/errored during screen-off: ${vm.error.value}",
            MonitorState.Monitoring,
            vm.state.value,
        )
        assertNull(vm.error.value)
        assertTrue(
            "no motion event recorded after the screen-off window (camera likely stalled)",
            secondEvent,
        )

        vm.stop()
        awaitState(MonitorState.Idle)
    }
}
